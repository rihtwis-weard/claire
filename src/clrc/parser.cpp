#include <iostream>

#include "parser.hpp"

namespace claire {

  auto Parser::next_state(ParseState prev, Token const &token) {
    return static_cast<ParseState>(parse_trans[utype(prev) + utype(token.kind)]);
  }

  std::unique_ptr<ASTNode> Parser::parse(std::vector<Token> const &tokens) {
    auto root  = std::make_unique<ASTNode>();
    auto state = ParseState::eNewScope;

    //      IO.puts("Hello world!")
    for (auto tok = tokens.begin(); tok != tokens.end(); tok++) {

      if (state = next_state(state, *tok); state <= ParseState::eFinal) {
        return root;
      }

      switch (state) {
      case ParseState::eExpr: {
        auto expr = parse_expr(state, tok);
        std::cout << "Expression: | " << tok->repr << " |" << std::endl;
        break;
      }
      default:
        break;
      }
    }

    return root;
  }

  std::unique_ptr<Expr> Parser::parse_expr(
    ParseState &state, std::vector<Token>::const_iterator tok) {
    auto node = std::make_unique<Expr>(tok->repr);

    for (; state > ParseState::eFinal; state = next_state(state, *(++tok))) {
      if (state = next_state(state, *tok); state <= ParseState::eFinal) {
        return node;
      }

      switch (state) {

      default:
        break;
      }
    }

    return node;
  }

} // namespace claire
