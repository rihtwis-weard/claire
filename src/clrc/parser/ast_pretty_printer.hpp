#pragma once

#include "ast.hpp"

namespace claire::parser {

  class ASTPrettyPrinter : ASTVisitor<std::string> {

  public:
    std::string pretty_print(ASTNode *node) {
      auto repr = indent(node->level());
      auto name = std::visit(*this, node->as_variant());
      repr += "└──" + name + "\n";

      for (auto const &child : node->children()) {
        repr += pretty_print(child.get());
      }
      return repr;
    }

    static std::string indent(std::size_t level) {
      std::string s{};
      for (std::size_t i = 0; i < level; i++) {
        s += "\t";
      }
      return s;
    }

    std::string operator()(ASTNode const *) override {
      return "ASTNode: root";
    }

    std::string operator()(ProgramDecl const *) override {
      return "ProgramDecl";
    }

    std::string operator()(ModuleDecl const *decl) override {
      return "ModuleDecl: " + decl->id();
    }

    std::string operator()(StringExpr const *expr) override {
      return "StringLiteral: " + expr->id();
    }

    std::string operator()(ExternDecl const *decl) override {
      return "ExternDecl: " + decl->id();
    }

    std::string operator()(IdentifierExpr const *expr) override {
      return expr->id();
    }

    std::string operator()(ModuleAccessExpr const *expr) override {
      return expr->module_name() + "." + expr->id();
    }

    std::string operator()(FunctionCallExpr const *expr) override {
      //      return "FunctionCallExpr: " + std::visit(*this, expr->callee()->as_variant());
      return "FunctionCallExpr";
    }

    std::string operator()(FunctionDef const *decl) override {
      return "FunctionDef: " + decl->id();
    }

    std::string operator()(FunctionBody const *decl) override {
      return "FunctionBody";
    }
  };

} // namespace claire::parser
