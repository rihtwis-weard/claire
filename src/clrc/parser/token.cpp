#include "token.h"

namespace claire::parser {
  robin_hood::unordered_map<std::string, TokenKind> const token_map = {
    // Operators
    {".", TokenKind::eAccess},
    {"(", TokenKind::eLParens},
    {"-", TokenKind::eMinus},
    // Special Multi Operators
    {"->", TokenKind::eArrow},
    {"|>", TokenKind::ePipe},
    // Reserved Keywords
    {"func", TokenKind::eReservedFunc},
    //    {"let", TokenKind::eReservedLet},
    //    {"if", TokenKind::eReservedIf},
    //    {"else", TokenKind::eReservedElse},
    //    {"open", TokenKind::eReservedOpen},
    //    {"module", TokenKind::eReservedModule},
    //    {"export", TokenKind::eReservedExport},
    //    {"extern", TokenKind::eReservedExtern},
    // Builtin Types
    {"binary", TokenKind::eTypeBinary},
    {"u32", TokenKind::eTypeU32},
  };
}
