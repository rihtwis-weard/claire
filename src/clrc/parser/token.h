#pragma once

#ifdef __cplusplus

#include <iomanip>

#include <robin_hood.h>

#include "../utils.hpp"

namespace claire::parser {

  enum class TokenKind {
    // General Tokens
    eIdentifier,
    eStringLiteral,
    eNumeral,
    eSeparator,
    eOperator,

    // Special Operators
    eAccessMember,
    eLParens,
    eRParens,
    eMinus,
    eScopeBegin,
    eScopeEnd,

    // Multi
    eAccessNamespace,
    eArrow,
    ePipe,

    // Reserved Keywords
    eReservedFunc,
    //    eReservedLet,
    //    eReservedIf,
    //    eReservedElse,
    //    eReservedOpen,
    //    eReservedModule,
    //    eReservedExport,
    //    eReservedExtern,

    // Builtin Types
    eTypeBinary,
    eTypeU32,

    eCount,
  };

  extern robin_hood::unordered_map<std::string, TokenKind> const token_map;

  struct Token {
    TokenKind   kind;
    std::string repr;
    std::size_t len;

    // Debugging metadata
//    std::string name;
    std::size_t line_num = 1;
    std::size_t col_num  = 1;

  public:
    auto to_hyponym() {
      if (auto it = token_map.find(repr); it != token_map.end()) {
        auto [_, token_kind] = *it;
        kind                 = token_kind;
      }
    }

    auto to_hyponym(TokenKind fallback) {
      if (auto it = token_map.find(repr); it != token_map.end()) {
        auto [_, token_kind] = *it;
        kind                 = token_kind;
      } else {
        kind = fallback;
      }
    }

    auto update_repr(char *src_ptr, int start_offset, int end_offset) {
      auto start = src_ptr - (len + start_offset);
      auto end   = start + (len + end_offset);
      repr       = std::string{start, end};
    }

    auto reset(std::size_t offset = 0) {
      // Reset trackers for next token
      col_num += (len + offset);
      len = 0;
    }

    friend std::ostream &operator<<(std::ostream &os, Token const &tok);
  };

#define TOKEN_DESC(kind, desc)                                                           \
  case (kind):                                                                           \
    os << (desc);                                                                        \
    break

  inline std::ostream &operator<<(std::ostream &os, Token const &tok) {
    os << std::setw(12);

    switch (tok.kind) {
      TOKEN_DESC(TokenKind::eIdentifier, "Identifier");
      TOKEN_DESC(TokenKind::eStringLiteral, "String");
      TOKEN_DESC(TokenKind::eNumeral, "Numeral");
      TOKEN_DESC(TokenKind::eSeparator, "Separator");
      TOKEN_DESC(TokenKind::eOperator, "Operator");
      TOKEN_DESC(TokenKind::eAccessMember, "Operator");
      TOKEN_DESC(TokenKind::eLParens, "Operator");
      TOKEN_DESC(TokenKind::eRParens, "Operator");
      TOKEN_DESC(TokenKind::eMinus, "Operator");
      TOKEN_DESC(TokenKind::eScopeBegin, "Separator");
      TOKEN_DESC(TokenKind::eScopeEnd, "Separator");
      TOKEN_DESC(TokenKind::eAccessNamespace, "Operator");
      TOKEN_DESC(TokenKind::eArrow, "Operator");
      TOKEN_DESC(TokenKind::ePipe, "Operator");

      TOKEN_DESC(TokenKind::eReservedFunc, "Keyword");
      //      TOKEN_DESC(TokenKind::eReservedLet, "Keyword");
      //      TOKEN_DESC(TokenKind::eReservedIf, "Keyword");
      //      TOKEN_DESC(TokenKind::eReservedElse, "Keyword");
      //      TOKEN_DESC(TokenKind::eReservedOpen, "Keyword");
      //      TOKEN_DESC(TokenKind::eReservedModule, "Keyword");
      //      TOKEN_DESC(TokenKind::eReservedExport, "Keyword");
      //      TOKEN_DESC(TokenKind::eReservedExtern, "Keyword");
      TOKEN_DESC(TokenKind::eTypeBinary, "Type");
      TOKEN_DESC(TokenKind::eTypeU32, "Type");
    default:
      os << "Unknown";
      break;
    }

    return os << " | " << tok.repr << " |";
  }

} // namespace claire::parser

#else

typedef enum {
  eTokenKindIdentifier,
  eTokenKindStringLiteral,
  eTokenKindNumeral,
  eTokenKindSeparator,
  eTokenKindOperator,
  eTokenKindAccessMember,
  eTokenKindLParens,
  eTokenKindRParens,
  eTokenKindMinus,
  eTokenKindScopeBegin,
  eTokenKindScopeEnd,
  eTokenKindAccessNamespace,
  eTokenKindArrow,
  eTokenKindPipe,
  eTokenKindReservedFunc,
  //  eTokenKindReservedLet,
  //  eTokenKindReservedIf,
  //  eTokenKindReservedElse,
  //  eTokenKindReservedOpen,
  //  eTokenKindReservedModule,
  //  eTokenKindReservedExport,
  //  eTokenKindReservedExtern,
  eTokenKindTypeBinary,
  eTokenKindTypeU32,
  eTokenKindCount,
} TokenKind;

#endif
