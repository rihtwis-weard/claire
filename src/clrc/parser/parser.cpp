#include <iostream>

#include "lexer.hpp"
#include "parser.hpp"

namespace claire::parser {

  auto Parser::next_state(ParseState prev, Token const &token) {
    return static_cast<ParseState>(
      parse_trans[utype(prev) + (utype(token.kind) * utype(ParseState::eCount))]);
  }

  template <typename RootNodeType>
  std::unique_ptr<ASTNode> Parser::parse(
    std::vector<Token> const &tokens, std::string const &id) {
    auto root  = std::make_unique<RootNodeType>(id);
    auto token = tokens.begin();

    switch (token->kind) {
    case TokenKind::eReservedFunc:
      root->add(parse_function_def(token));
      break;
    default:
      break;
    }

    return root;
  }

  std::unique_ptr<FunctionDef> Parser::parse_function_def(
    std::vector<Token>::const_iterator &tok) {
    tok              = std::next(tok);
    auto id          = tok->repr;
    auto symbol_name = std::make_unique<IdentifierExpr>(tok->repr);

    tok       = std::next(tok);
    auto args = parse_function_decl_args(tok);

    // TODO(rw): parse return type decl

    tok = std::next(tok);
    if (tok->kind == TokenKind::eScopeBegin) {
      auto body = parse_function_body(tok);
      return std::make_unique<FunctionDef>(
        id, std::move(symbol_name), std::move(args), std::move(body));
    } else {
      // TODO(rw): parse semicolon for func decl
      throw syntax_error{};
    }
  }

  std::unique_ptr<FunctionBody> Parser::parse_function_body(
    std::vector<Token>::const_iterator &tok) {
    tok       = std::next(tok);
    auto body = std::make_unique<FunctionBody>();

    for (; tok->kind != TokenKind::eScopeEnd;) {
      switch (tok->kind) {
      case TokenKind::eIdentifier:
        body->add(parse_identifier_expr(tok));
        break;
      case TokenKind::eLParens:
        body->add(parse_function_call_expr(tok));
        break;
      default:
        throw syntax_error{};
      }
    }

    return body;
  }

  std::unique_ptr<IdentifierExpr> Parser::parse_identifier_expr(
    std::vector<Token>::const_iterator &tok) {
    // TODO(rw): handle fully qualified names and field/method access

    auto ident = std::make_unique<IdentifierExpr>(tok->repr);
    tok        = std::next(tok);

    // TODO(rw): clean up breaking out of loop
    auto is_identifier_seq = [](auto kind) -> bool {
      return kind == TokenKind::eAccessNamespace or kind == TokenKind::eIdentifier;
    };

    for (; is_identifier_seq(tok->kind); ++tok) {
      switch (tok->kind) {
      case TokenKind::eAccessNamespace:
        tok = std::next(tok);
        ident->add(parse_identifier_expr(tok));
        break;
      default:
        break;
      }
    }
    return ident;
  }

  //  template <typename ASTNodeType>
  //  std::unique_ptr<ASTNode> Parser::parse(
  //    std::vector<Token> const &tokens, std::string const &id) {
  //    auto root  = std::make_unique<ASTNodeType>(id);
  //    // TODO(rihtwis-weard): `NewScope` synonymous with an additional `Program` state?
  //    auto state = ParseState::eNewScope;
  //
  //    for (auto tok = tokens.begin(); tok != tokens.end(); tok++) {
  //      if (state = next_state(state, *tok); state <= ParseState::eFinal) {
  //        return root;
  //      }
  //
  //      switch (state) {
  //      case ParseState::eIdentifierExpr: {
  //        root->add(parse_identifier_expr(state, tok));
  //        break;
  //      }
  //      case ParseState::eModuleOpenStmt: {
  //        // TODO(rihtwis-weard): elide `open ModuleName` declaration for stdlib modules
  //        root->add(parse_module_open_stmt(state, tok, tokens));
  //        state = ParseState::eNewScope;
  //        break;
  //      }
  //      case ParseState::eExternDecl: {
  //        root->add(parse_extern_decl(state, tok, tokens));
  //        break;
  //      }
  //      default:
  //        break;
  //      }
  //    }
  //
  //    return root;
  //  }

  template std::unique_ptr<ASTNode> Parser::parse<ProgramDecl>(
    std::vector<Token> const &tokens, std::string const &id);

  template std::unique_ptr<ASTNode> Parser::parse<ModuleDecl>(
    std::vector<Token> const &tokens, std::string const &id);

  //  std::unique_ptr<Expr> Parser::parse_identifier_expr(
  //    ParseState &state, std::vector<Token>::const_iterator tok) {
  //    std::unique_ptr<IdentifierExpr> node = std::make_unique<IdentifierExpr>(tok->repr);
  //
  //    for (; state > ParseState::eFinal;) {
  //      if (state = next_state(state, *(++tok)); state <= ParseState::eFinal) {
  //        return node;
  //      }
  //
  //      switch (state) {
  //      case ParseState::eNewAccessExpr: {
  //        return parse_module_access_expr(tok, *node);
  //      }
  //      default:
  //        break;
  //      }
  //    }
  //
  //    return node;
  //  }

  //  std::unique_ptr<Expr> Parser::parse_module_access_expr(
  //    std::vector<Token>::const_iterator tok, IdentifierExpr const &expr) {
  //
  //    auto node = std::make_unique<ModuleAccessExpr>(expr);
  //
  //    for (auto state = ParseState::eNewAccessExpr; state > ParseState::eFinal;) {
  //      if (state = next_state(state, *(++tok)); state <= ParseState::eFinal) {
  //        return node;
  //      }
  //
  //      if (state == ParseState::eGrowAccessExpr) {
  //        node->grow(tok->repr);
  //      } else if (state == ParseState::eFunctionCallExpr) {
  //        return parse_function_call_expr(tok, std::move(node));
  //      }
  //    }
  //    return node;
  //  }

  std::unique_ptr<Expr> Parser::parse_function_call_expr(
    std::vector<Token>::const_iterator &tok) {
    auto node = std::make_unique<FunctionCallExpr>();

    // TODO(rw): distinguish right parens and separator
    for (; tok->kind != TokenKind::eSeparator; ++tok) {
      switch (tok->kind) {
        // TODO(rihwis-weard): don't hardcode type / values
      case TokenKind::eStringLiteral:
        node->add(std::make_unique<StringExpr>(tok->repr));
        break;
      default:
        break;
      }
    }

    return node;
  }

  //  std::unique_ptr<Expr> Parser::parse_function_call_expr(
  //    std::vector<Token>::const_iterator tok, std::unique_ptr<Expr> &&callee) {
  //    auto node = std::make_unique<FunctionCallExpr>(std::move(callee));
  //
  //    for (auto state = ParseState::eFunctionCallExpr; state > ParseState::eFinal;) {
  //      if (state = next_state(state, *(++tok)); state <= ParseState::eFinal) {
  //        return node;
  //      }
  //
  //      // TODO(rihwis-weard): don't hardcode type / values
  //      if (state == ParseState::eFunctionCallArgs) {
  //        node->add(std::make_unique<StringExpr>(tok->repr));
  //      }
  //    }
  //
  //    return node;
  //  }

  std::unique_ptr<ASTNode> Parser::parse_module_open_stmt(ParseState &state,
    std::vector<Token>::const_iterator &tok, std::vector<Token> const &tokens) {

    auto module_root = std::make_unique<ASTNode>();

    for (; state > ParseState::eFinal;) {
      if (++tok == tokens.end()) {
        return module_root;
      }

      if (state = next_state(state, *tok); state <= ParseState::eFinal) {
        return module_root;
      }

      if (state == ParseState::eIdentifierExpr) {
        // TODO(rihwis-weard): correct way to cache repeated ASTNode/Module resolutions and returning
        //                     unique_ptr to ASTNode? May need to use shared_ptr/weak_ptrs all around
        //        if (auto mod = mod_map_.find(tok->repr); mod != mod_map_.end()) {
        //          return std::unique_ptr<ASTNode>{mod->second.get()};
        //        }

        // TODO(rihtwis-weard): return `OpenModuleStmt` if `Module` is already loaded

        // TODO(rihwis-weard): don't hard-code module lookup/resolution
        // distribute module registry, file -> module names
        auto toks = Lexer{stdlib_path_ + "io.clr"}.lex();

        // TODO(rihtwis-weard): module not found exception or save error and keep parsing
        return parse<ModuleDecl>(toks, tok->repr);
      }
    }

    return module_root;
  }

  std::unique_ptr<ExternDecl> Parser::parse_extern_decl(ParseState &state,
    std::vector<Token>::const_iterator tok, std::vector<Token> const &tokens) {

    std::string              name{};
    std::vector<FunctionArg> args{};
    std::string              return_type{};
    std::string              linkage_name{};

    for (; state > ParseState::eFinal;) {
      if (++tok == tokens.end()) {
        return std::make_unique<ExternDecl>(
          name, std::move(args), return_type, std::make_unique<StringExpr>(linkage_name));
      }

      if (state = next_state(state, *tok); state <= ParseState::eFinal) {
        return std::make_unique<ExternDecl>(
          name, std::move(args), return_type, std::make_unique<StringExpr>(linkage_name));
      }

      if (state == ParseState::eExternDeclName) {
        name = tok->repr;
      } else if (state == ParseState::eFunctionDeclArgs) {
        args = parse_function_decl_args(tok);
        --tok;
      } else if (state == ParseState::eFunctionDeclReturnType) {
        tok++;
        return_type = tok->repr;
        state       = ParseState::eExternDecl;
      } else if (state == ParseState::eExternDeclLinkageName) {
        linkage_name = tok->repr;
      }
    }

    // TODO(rihtwis-weard): need to record errors
    return nullptr;
  }

  std::vector<FunctionArg> Parser::parse_function_decl_args(
    std::vector<Token>::const_iterator &tok) {
    if (tok->kind != TokenKind::eLParens) {
      throw syntax_error{};
    }

    auto args = std::vector<FunctionArg>{};

    FunctionArg curr_arg{};
    tok = std::next(tok);

    // TODO(rw): distinguish right parens and operator
    for (; tok->kind != TokenKind::eOperator; ++tok) {
      tok           = std::next(tok);
      curr_arg.name = tok->repr;

      // TODO(rw): parse colon separator, type identifier, and comma separator

      // args.push_back(curr_arg);
    }
    return args;
  }

} // namespace claire::parser
