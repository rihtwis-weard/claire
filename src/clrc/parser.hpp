#pragma once

#include <vector>

#include "ast.hpp"
#include "state_machine.h"
#include "token.h"

namespace claire {

  class Parser {
  public:
    std::unique_ptr<ASTNode> parse(std::vector<Token> const &tokens);

  private:
    std::unique_ptr<Expr> parse_expr(
      ParseState &state, std::vector<Token>::const_iterator tok);
    std::unique_ptr<Expr> parse_call_expr(std::unique_ptr<Expr> &&fn);

    auto next_state(ParseState prev, Token const &token);
  };
} // namespace claire
