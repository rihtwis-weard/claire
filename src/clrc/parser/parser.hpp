#pragma once

#include <utility>
#include <vector>

#include <robin_hood.h>

#include "ast.hpp"
#include "state_machine.h"
#include "token.h"

namespace claire::parser {

  using token_iterator = std::vector<Token>::const_iterator;

  struct parse_context {
    std::vector<Token> const &tokens;

    token_iterator tok;

  public:
    explicit parse_context(std::vector<Token> const &tokens)
      : tokens{tokens}
      , tok{tokens.begin()} {
    }
  };

  std::unique_ptr<IdentifierExpr> parse_simple_identifier_expression(token_iterator &tok);

  std::unique_ptr<IdentifierSeq> parse_identifier_sequence(token_iterator &tok);

  std::unique_ptr<FunctionCallExpr> parse_function_call_expression(
    parse_context &ctx, std::unique_ptr<Expr> &&callee);

  std::unique_ptr<ExpressionSequence> parse_expression_sequence(parse_context &ctx);

  class Parser {
    std::string stdlib_path_;
    //    robin_hood::unordered_map<std::string, ASTNode *> mod_map_;

  public:
    explicit Parser(std::string stdlib_path)
      : stdlib_path_{std::move(stdlib_path)} {
    }

    template <typename RootNodeType = ProgramDecl>
    std::unique_ptr<ASTNode> parse(
      std::vector<Token> const &tokens, std::string const &id = "main");

  private:
    //    std::unique_ptr<Expr> parse_module_access_expr(
    //      std::vector<Token>::const_iterator tok, IdentifierExpr const &expr);

    //
    //    std::unique_ptr<ASTNode> parse_module_open_stmt(ParseState &state,
    //      std::vector<Token>::const_iterator &tok, std::vector<Token> const &tokens);
    //
    //    std::unique_ptr<ExternDecl> parse_extern_decl(ParseState &state,
    //      std::vector<Token>::const_iterator tok, std::vector<Token> const &tokens);

    std::unique_ptr<FunctionDef> parse_function_def(
      std::vector<Token>::const_iterator &tok);

    std::unique_ptr<FunctionBody> parse_function_body(
      std::vector<Token>::const_iterator &tok);
  };

} // namespace claire::parser
