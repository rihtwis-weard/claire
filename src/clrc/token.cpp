#include "token.hpp"

namespace claire {
  robin_hood::unordered_map<std::string, TokenKind> const token_map = {
    // Special Multi Operators
    {"|>", TokenKind::ePipe},
    // Reserved Keywords
    {"let", TokenKind::eReservedLet},
    {"if", TokenKind::eReservedIf},
    {"else", TokenKind::eReservedElse},
  };
}
