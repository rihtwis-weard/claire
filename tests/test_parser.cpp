#include "fixtures.hpp"
#include "parser/ast_pretty_printer.hpp"
#include "parser/parser.hpp"
#include "parser/token.h"

int main() {

  "simple_identifier"_test = []() {
    auto pp = claire::parser::ASTPrettyPrinter{};

    std::vector<claire::parser::Token> const tokens{
      {
        .kind = claire::parser::TokenKind::eIdentifier,
        .repr = "my_variable",
        .len  = 11,
      },
    };

    auto it   = tokens.begin();
    auto node = claire::parser::parse_simple_identifier_expression(it);
    Approvals::verify(pp.pretty_print(node.get()));
  };
}
