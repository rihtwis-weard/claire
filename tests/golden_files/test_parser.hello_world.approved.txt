└──ProgramDecl
	└──FunctionCall: AccessExpression: IO.puts
		└──StringLiteral: "Hello world!"

